foundations/foundation-metrics/src/main/java/io/servicecomb/foundation/metrics/performance/MetricsDataMonitorUtil.java

```diff
@@ -0,0 +1,236 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.servicecomb.foundation.metrics.performance;
+
+import io.servicecomb.core.Invocation;
+import io.servicecomb.core.QueueMetricTimes;
+import io.servicecomb.swagger.invocation.InvocationType;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of metric util functions such as preparing average queue times
+ * and total requests/fail for producer and instance level.
+ */
+public final class MetricsDataMonitorUtil {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MetricsDataMonitorUtil.class);
+
+  /**
+   * start time in queue.
+   */
+  public static void prepareStartTimeMetrics(Invocation invocation) {
+    // Notedown starting time here in queue.
+    if ((invocation == null) || (invocation.getOperationMeta() == null)) {
+      LOGGER.error("Invocation is not correct");
+      return;
+    }
+
+    String operPath = invocation.getOperationMeta().getMicroserviceQualifiedName();
+    MetricsDataMonitor.getMetricsGlobalRef().incrementQueueCount();
+
+    QueueMetricTimes queueMetricTimes = invocation.getQueueMetricTimes();
+    if (queueMetricTimes == null) {
+      queueMetricTimes = new QueueMetricTimes();
+      invocation.setQueueMetricTimes(queueMetricTimes);
+    }
+
+    QueueMetricsData reqQueue = MetricsDataMonitor.getMetricsGlobalRef().getOrCreateOperationMetrics(operPath);
+    queueMetricTimes.setQueueStartTime(System.currentTimeMillis());
+    reqQueue.incrementCountInQueue();
+  }
+
+  /**
+   * Prepare the queue time metrics.
+   */
+  public static void prepareQueueTimeMetrics(Invocation invocation) {
+
+    if ((invocation == null) || (invocation.getOperationMeta() == null)) {
+      LOGGER.error("Invocation is not correct");
+      return;
+    }
+
+    String operPath = invocation.getOperationMeta().getMicroserviceQualifiedName();
+    QueueMetricTimes queueMetricTimes = invocation.getQueueMetricTimes();
+    if (queueMetricTimes == null) {
+      LOGGER.error("Queue time metrics can not be NULL");
+      return;
+    }
+
+    queueMetricTimes.setQueueEndtime(System.currentTimeMillis());
+    QueueMetricsData reqQueue = MetricsDataMonitor.getMetricsGlobalRef().getOrCreateOperationMetrics(operPath);
+    reqQueue.decrementCountInQueue();
+    MetricsDataMonitor.getMetricsGlobalRef().decrementQueueCount();
+
+    // calculate average time in queue.
+    long timeInqueue = invocation.getQueueMetricTimes().getQueueEndtime()
+        - invocation.getQueueMetricTimes().getQueueStartTime();
+    reqQueue.getTotalTime().addAndGet(timeInqueue);
+    reqQueue.incrementTotalCount();
+
+    double avgTimeInQueue = reqQueue.getTotalTime().get() / reqQueue.getTotalCount().get();
+
+    // store average,min and max times in operation level first.
+    QueueMetrics requestMetricQueue = MetricsDataMonitor.getMetricsGlobalRef().getQueueMetrics(operPath);
```

Comment:
concurrence problem?

and minTime/maxTime also have the problem?

Reply:
fixed